Firestore/core/src/firebase/firestore/nanopb/byte_string.h

```diff
@@ -71,34 +93,60 @@ class ByteString : public util::Comparable<ByteString> {
   }
 
   /**
-   * Creates a new ByteString that takes ownership of the given byte array.
+   * Creates a new ByteString that copies the contents of the given byte array.
+   */
+  static ByteString Copy(const pb_bytes_array_t* bytes);
+
+  /**
+   * Creates a new ByteString that takes ownership of the given byte array. If
+   * taking from a nanopb-created message struct, the caller should null out
+   * the pointer there so that pb_release won't free the buffer that the
+   * returned ByteString now owns.
    */
   static ByteString Take(pb_bytes_array_t* bytes);
 
   /**
-   * Returns a pointer to the character data backing this ByteString. The return
-   * value is `nullptr` if the backing bytes are themselves null.
+   * Returns a pointer to the character data backing this ByteString. The
+   * returned buffer is always non-null, even if the nanopb byte array is null.
    */
-  const uint8_t* data() const {
-    return bytes_ ? bytes_->bytes : nullptr;
-  }
+  const uint8_t* data() const;
 
   size_t size() const {
     return bytes_ ? bytes_->size : 0;
   }
 
   bool empty() const {
-    return size() == 0;
+    return bytes_ == nullptr || bytes_->size == 0;
+  }
+
+  const uint8_t* begin() const {
+    return data();
   }
 
-  /** Returns a const view of the underlying byte array. */
+  const uint8_t* end() const {
+    return data() + size();
+  }
+
+  /**
+   * Returns a const view of the raw underlying byte array pointer.
+   *
+   * This value may be null because nanopb (and protobuf generally) treat null
+   * and empty byte arrays as equivalent.
+   *
+   * For actually reading the data in the buffer, prefer `data()` and `size()`
+   * or `begin()` and `end()`, which handle this nullability for you.
+   */
   const pb_bytes_array_t* get() const {
     return bytes_;
   }
 
   /**
-   * Returns the current byte array and assigns the backing byte array to
-   * nullptr, releasing the ownership of the array contents to the caller.
+   * Releases ownership of the backing byte array, and returns it to the caller.
+   * The backing byte array is set to null.
+   *
+   * This value may be null because nanopb (and protobuf generally) treat null
```

Comment:
Question: I might have missed something, but all the uses of `release` I remember from reading this PR were of the form:
```cpp
nanopb_type.bytes = ByteString{some_input).release(); // avoid double-free
```
In other words, `ByteString` was used as a way to create a deep copy of the pointer. This pattern seems like it's almost creating an object for a side effect, not to actually use it.

I don't know if there are other known uses of `release` in the future (or I missed something in this PR). If not, then it seems like providing a dedicated copy function would not only allow eliminating `release`, but perhaps allow changing the internal representation to `vector<uint8_t>`? On the other hand, it would make `Take` impossible -- I don't know how much of a performance hit that would be.